decoder/src/lib.rs

```diff
@@ -98,13 +98,69 @@ pub struct Table {
 /// Checks if the version encoded in the symbol table is compatible with this version of the
 /// `decoder` crate
 pub fn check_version(version: &str) -> Result<(), String> {
+    enum Kind {
+        // "1" or "0.1"
+        Semver,
+        // commit hash "e739d0ac703dfa629a159be329e8c62a1c3ed206"
+        Git,
+    }
+
+    impl Kind {
+        fn of(version: &str) -> Kind {
+            if version.contains('.') {
+                // "0.1"
+                Kind::Semver
+            } else if version.parse::<u64>().is_ok() {
+                // "1"
+                Kind::Semver
+            } else {
+                // "e739d0ac703dfa629a159be329e8c62a1c3ed206" (should be)
+                Kind::Git
+            }
+        }
+    }
+
     if version != DEFMT_VERSION {
-        return Err(format!(
-            "defmt version mismatch (firmware is using {}, host is using {}); \
-             are you using the same git version of defmt and related tools?
-Try `cargo install`-ing the latest git version of `probe-run` AND updating your project's `Cargo.lock` with `cargo update`",
-            version, DEFMT_VERSION,
-        ));
+        let mut msg = format!(
+            "defmt version mismatch: firmware is using {}, `probe-run` supports {}\nsuggestion: ",
+            version, DEFMT_VERSION
+        );
+
+        match (Kind::of(version), Kind::of(DEFMT_VERSION)) {
+            (Kind::Git, Kind::Git) => {
+                write!(
+                    msg,
+                    "pin _all_ `defmt` related dependencies to revision {0}; modify Cargo.toml files as shown below
+
+[dependencies]
+defmt = {{ git = \"https://github.com/knurling-rs/defmt\", rev = \"{0}\" }}
+defmt-rtt = {{ git = \"https://github.com/knurling-rs/defmt\", rev = \"{0}\" }}
+# ONLY pin this dependency if you are using the `print-defmt` feature
+panic-probe = {{ git = \"https://github.com/knurling-rs/defmt\", features = [\"print-defmt\"], rev = \"{0}\" }}",
```

Comment:
`panic-probe` currently lives in the `probe-run` repo so it's not really possible to pin to the correct defmt rev (`panic-probe` also already pins to a particular `defmt` and that cannot be changed from an application project). I think that to simplify the use of the git-version of defmt we'll have to move panic-probe (back) into the defmt repo. This PR is blocked on that migration.

Reply:
migration PR: #244